domain/blockdag/utxoset.go

```diff
@@ -488,13 +509,83 @@ func (fus *FullUTXOSet) containsInputs(tx *appmessage.MsgTx) bool {
 
 // clone returns a clone of this utxoSet
 func (fus *FullUTXOSet) clone() UTXOSet {
-	return &FullUTXOSet{utxoCollection: fus.utxoCollection.clone()}
+	return &FullUTXOSet{
+		utxoCollection:   fus.utxoCollection.clone(),
+		dbContext:        fus.dbContext,
+		size:             fus.size,
+		maxUTXOCacheSize: fus.maxUTXOCacheSize,
+	}
+}
+
+// this method hides the embedded utxoCollection's implementation of get
+func (fus *FullUTXOSet) get(outpoint appmessage.Outpoint) (*UTXOEntry, bool) {
+	return fus.Get(outpoint)
+}
+
+// getSizeOfUTXOEntryAndOutpoint returns size of UTXOEntry & Outpoint in bytes
+func getSizeOfUTXOEntryAndOutpoint(entry *UTXOEntry) int64 {
+	const staticSize = int64(unsafe.Sizeof(UTXOEntry{}) + unsafe.Sizeof(appmessage.Outpoint{}))
+	return staticSize + int64(len(entry.scriptPubKey))
+}
+
+// checkAndCleanCachedData checks the FullUTXOSet size and clean it if it reaches the limit
+func (fus *FullUTXOSet) checkAndCleanCachedData() {
+	if fus.maxUTXOCacheSize > 0 && fus.size > fus.maxUTXOCacheSize {
+		fus.utxoCollection = make(utxoCollection)
+		fus.size = 0
+	}
+}
+
+// add adds a new UTXO entry to this FullUTXOSet
+// this method hides the embedded utxoCollection's implementation of add
+func (fus *FullUTXOSet) add(outpoint appmessage.Outpoint, entry *UTXOEntry) {
+	fus.utxoCollection[outpoint] = entry
+	fus.size += getSizeOfUTXOEntryAndOutpoint(entry)
+	fus.checkAndCleanCachedData()
+}
+
+// remove removes a UTXO entry from this collection if it exists
+// this method hides the embedded utxoCollection's implementation of remove
+func (fus *FullUTXOSet) remove(outpoint appmessage.Outpoint) {
+	entry, ok := fus.utxoCollection.get(outpoint)
+	if ok {
+		delete(fus.utxoCollection, outpoint)
+		fus.size -= getSizeOfUTXOEntryAndOutpoint(entry)
+	}
 }
 
 // Get returns the UTXOEntry associated with the given Outpoint, and a boolean indicating if such entry was found
+// If the UTXOEntry is not exist in the memory then check in the database
 func (fus *FullUTXOSet) Get(outpoint appmessage.Outpoint) (*UTXOEntry, bool) {
 	utxoEntry, ok := fus.utxoCollection[outpoint]
-	return utxoEntry, ok
+	if ok {
+		return utxoEntry, ok
+	}
+
+	if fus.dbContext == nil {
```

Comment:
This is to make NewFullUTXOSet structure elastic, and valid if it was initialized without context.
Also, there are unit tests that test the fullUTXO without dbContext.
This is a controversial decision, so it has both downsides and upsides.

Reply:
We don't want to support this kind of functionality, and I don't want to test one thing on unit test and another thing on productions. So please make remove this condition, and then if you won't fill it, it'll panic because of nil dereference.